server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,443 @@
+        writeLock.lock();
+        try {
+            return localDirectory.createTempOutput(prefix, suffix, context);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Ensures that any writes to these files are moved to stable storage (made durable).
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void sync(Collection<String> names) throws IOException {
+        logger.trace("sync() called {}", names);
+        writeLock.lock();
+        try {
+            Collection<String> newLocalFiles = new ArrayList<>();
+            Collection<String> remoteFiles = Arrays.asList(getRemoteFiles());
+            for (String name : names) {
+                if (remoteFiles.contains(name) == false) newLocalFiles.add(name);
+            }
+            logger.trace("Synced files : {}", newLocalFiles);
+            localDirectory.sync(newLocalFiles);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Ensures that directory metadata, such as recent file renames, are moved to stable storage.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void syncMetaData() throws IOException {
+        logger.trace("syncMetaData() called ");
+        writeLock.lock();
+        try {
+            localDirectory.syncMetaData();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Renames {@code source} file to {@code dest} file where {@code dest} must not already exist in
+     * the directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        logger.trace("rename() called {}, {}", source, dest);
+        writeLock.lock();
+        try {
+            localDirectory.rename(source, dest);
+            fileCache.remove(localDirectory.getDirectory().resolve(source));
+            cacheFile(dest);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Opens a stream for reading an existing file.
+     * Check whether the file is present locally or in remote and return the IndexInput accordingly
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        logger.trace("openInput() called {}", name);
+        writeLock.lock();
+        try {
+            remoteDirectoryBlobStreamReader.setContext(context);
+            /**
+             * We aren't tracking temporary files (created via createTempOutput) currently in FileCache as these are created and then deleted within a very short span of time
+             * We will be reading them directory from the local directory
+             */
+            if (isTempFile(name)) {
+                return localDirectory.openInput(name, context);
```

Comment:
Since the temp file naming controlled via the underlying FSDirectory implementation, it is not possible to track these files in a FileCache (as we would need their path for caching them). Not sure on the size of these files, will update after some investigation. But usually these files are deleted before they are even uploaded to the remote store.